### PR TITLE
Feature/add storage data to juju status

### PR DESCRIPTION
**Pull request by ycliuhw**
## Description of change

introduce storage information to status cmd;

## QA steps

1. deploy some workloads with storage;
2. `juju status --storage  -m <model-name>`;
3. `juju status --format json -m <model-name> | python -m json.tool`;
4. `juju status --format yaml -m <model-name> `;
5. `juju status --format yaml -m <model-name> --storage`;
6. `juju status --format yaml -m <model-name> --relations --storage`;

### current status example
```
Model    Controller  Cloud/Region         Version      SLA          Timestamp
default  ctrl1       localhost/localhost  2.5-beta2.1  unsupported  13:21:45+11:00

App         Version  Status   Scale  Charm       Store       Rev  OS      Notes
postgresql           waiting    0/1  postgresql  jujucharms  187  ubuntu  

Unit           Workload  Agent       Machine  Public address  Ports  Message
postgresql/1*  waiting   allocating  1        10.10.145.50           agent initializing

Machine  State    DNS           Inst id        Series  AZ  Message
1        started  10.10.145.50  juju-052c15-1  bionic      Running

Relation provider       Requirer                Interface    Type  Message
postgresql:coordinator  postgresql:coordinator  coordinator  peer  joining  
postgresql:replication  postgresql:replication  pgpeer       peer  joining  

Storage Unit  Storage id  Type        Pool  Mountpoint  Size    Status     Message
postgresql/1  pgdata/1    filesystem  lxd               100GiB  attaching  attaching filesystem 1 to machine 1: Missing device type for device 'filesystem-1'

```

## Documentation changes

Yes.

## Bug reference

https://bugs.launchpad.net/juju/+bug/1653753

### CODE REVIEW

**Comment by ycliuhw** [comment]
hi @wallyworld 
how about this format of storage status?

Model  Controller  Cloud/Region  Version      SLA          Timestamp
t1     ctrl1       microk8s3     2.5-beta2.1  unsupported  12:07:09+11:00

App          Version  Status  Scale  Charm        Store  Rev  OS          Address         Notes
gitlab-k8s            active      1  gitlab-k8s   local    0  kubernetes  10.152.183.218  
mariadb-k8s           active      1  mariadb-k8s  local    0  kubernetes  10.152.183.183  

Unit            Workload  Agent  Address     Ports     Message
gitlab-k8s/0*   active    idle   10.1.1.236  80/TCP    
mariadb-k8s/0*  active    idle   10.1.1.232  3306/TCP  

Relation provider   Requirer          Interface  Type     Message
mariadb-k8s:server  gitlab-k8s:mysql  mysql      regular  

Unit           Storage Id  Type        Pool        Size   Status    Message
mariadb-k8s/0  database/0  filesystem  mariadb-pv  38MiB  attached  

Unit           Storage Id  Filesystem Id  Provider id                         Mountpoint      Size   State     Message
mariadb-k8s/0  database/0  0              juju-database-0-juju-mariadb-k8s-0  /var/lib/mysql  38MiB  attached  

Unit           Storage Id  Volumes Id  Provider Id                               Size    State     Message
mariadb-k8s/0  database/0  0           pvc-8ae0057b-ec84-11e8-a504-9cb6d0fbc7cd  641MiB  attached

**Comment by ycliuhw** [comment]
[juju should surface blocked volume deletion better](https://trello.com/c/zLjRSvsn/542-juju-should-surface-blocked-volume-deletion-better)

File: cmd/juju/status/formatter.go [hunk 1]
```diff
@@ -14,34 +14,49 @@ import (
...
 	"github.com/juju/juju/state/multiwatcher"
 )
 
 type statusFormatter struct {
-	status         *params.FullStatus
-	controllerName string
-	relations      map[int]params.RelationStatus
-	isoTime        bool
-	showRelations  bool
+	storage                             *storage.CombinedStorage
```
Comment (wallyworld) [inline cmd/juju/status/formatter.go]: nit - can we move this new storage down under relations

File: cmd/juju/status/formatter.go [hunk 2]
```diff
@@ -14,34 +14,49 @@ import (
...
-		controllerName: controllerName,
+		storage:        p.storage,
+		status:         p.status,
+		controllerName: p.controllerName,
 		relations:      make(map[int]params.RelationStatus),
-		isoTime:        isoTime,
-		showRelations:  showRelations,
+		isoTime:        p.isoTime,
+		showRelations:  p.showRelations,
+		showStorage:    p.showStorage,
```
Comment (wallyworld) [inline cmd/juju/status/formatter.go]: We don't need showStorage on statusFormatter - if storage is nil, that means don't show it

File: cmd/juju/status/status.go [hunk 3]
```diff
@@ -176,23 +192,72 @@ func (c *statusCommand) Init(args []string) error {
 }
 
 var newAPIClientForStatus = func(c *statusCommand) (statusAPI, error) {
-	if c.api != nil {
-		return c.api, nil
+	if c.statusAPI == nil {
```
Comment (wallyworld) [inline cmd/juju/status/status.go]: Normally we do != nil and exit early

File: cmd/juju/status/status.go [hunk 4]
```diff
@@ -176,23 +192,72 @@ func (c *statusCommand) Init(args []string) error {
...
+		if err != nil {
+			return nil, err
+		}
+		c.storageAPI = storageapi.NewClient(root)
 	}
-	return c.NewAPIClient()
+	return c.storageAPI, nil
+}
+
+func (c *statusCommand) close() (err error) {
```
Comment (wallyworld) [inline cmd/juju/status/status.go]: This isn't right - regardless of where the api came from, we close it. Just use a defer inside Run()

File: cmd/juju/storage/volumelistformatters.go [hunk 5]
```diff
@@ -70,9 +69,9 @@ func formatVolumeListTabular(writer io.Writer, infos map[string]VolumeInfo) erro
 	sort.Sort(volumeAttachmentInfos)
 
 	if haveMachines {
-		print("Machine", "Unit", "Storage", "Id", "Provider Id", "Device", "Size", "State", "Message")
+		print("Machine", "Unit", "Storage Id", "Volumes Id", "Provider Id", "Device", "Size", "State", "Message")
```
Comment (wallyworld) [inline cmd/juju/storage/volumelistformatters.go]: Where we have "Storage Id", it needs to be "Storage id". Same for Volume Id etc. Also not Volumes
Looks like the original code here was wrong.

File: cmd/juju/storage/listformatters.go [hunk 6]
```diff
@@ -88,15 +90,76 @@ func formatStorageListTabular(
...
+
+// FormatStorageListForStatusTabular writes a tabular summary of storage for status tabular view.
+func FormatStorageListForStatusTabular(writer *ansiterm.TabWriter, s CombinedStorage) error {
+	w := output.Wrapper{writer}
+
+	storagePool, storageSize := getStoragePoolAndSize(s)
+	units, byUnit := sortStorageInstancesByUnitId(s)
+
+	w.Println()
+	w.Print("Unit", "Storage Id", "Type")
```
Comment (wallyworld) [inline cmd/juju/storage/listformatters.go]: Storage id

File: cmd/juju/storage/listformatters.go [hunk 7]
```diff
@@ -11,55 +11,57 @@ import (
...
-		// rather than the volume.
-		if _, ok := storageSize[v.Storage]; !ok {
-			storageSize[v.Storage] = v.Size
-		}
-	}
+	storagePool, storageSize := getStoragePoolAndSize(s)
+	units, byUnit := sortStorageInstancesByUnitId(s)
 
-	w.Print("Unit", "Id", "Type")
+	w.Print("Unit", "Storage Id", "Type")
```
Comment (wallyworld) [inline cmd/juju/storage/listformatters.go]: Storage id

File: cmd/juju/storage/filesystemlistformatters.go [hunk 8]
```diff
@@ -70,9 +69,9 @@ func formatFilesystemListTabular(writer io.Writer, infos map[string]FilesystemIn
 	sort.Sort(filesystemAttachmentInfos)
 
 	if haveMachines {
-		print("Machine", "Unit", "Storage", "Id", "Volume", "Provider id", "Mountpoint", "Size", "State", "Message")
+		print("Machine", "Unit", "Storage Id", "Id", "Volume", "Provider id", "Mountpoint", "Size", "State", "Message")
```
Comment (wallyworld) [inline cmd/juju/storage/filesystemlistformatters.go]: Storage id

File: cmd/juju/status/formatter.go (hunk 2 above)
Comment (ycliuhw) [reply cmd/juju/status/formatter.go]: I am thinking we probably still show an empty storage section if user specified `--storage` even if the storage is `nil`. And we should not show the storage section if the user didn't specify `--storage` even if the storage is not `nil`, right?

Comment (ycliuhw) [reply cmd/juju/status/formatter.go]: ah, you are right. thanks!

**Comment by ycliuhw** [comment]
$$merge$$

**Pull request merged**